File: iosApp/iosApp/SocialLinks/SocialLinkMapper.swift
//
//  SocialLinksMapper.swift
//  iosApp
//
//  Created by Yammine on 4/27/23.
//  Copyright © 2023 tomorrowSARL. All rights reserved.
//

import Foundation
import shared

extension Array where Element == SharedSocialLink {
    func mapToSocialLinkUI() -> [SocialLink] {
        SocialLink.allCases.map { emptyLink in
           let link = self.first { link in
               switch(emptyLink) {
               case .linkedIn(_): return link.platform == SharedSocialLink.Platform.linkedin
               case .twitter(_): return link.platform == SharedSocialLink.Platform.twitter
               case .website(_): return link.platform == SharedSocialLink.Platform.website
               case .instagram(_): return link.platform == SharedSocialLink.Platform.instagram
               case .facebook(_): return link.platform == SharedSocialLink.Platform.facebook
               case .whatsapp(_): return link.platform == SharedSocialLink.Platform.whatsapp
               case .phone(_): return link.platform == SharedSocialLink.Platform.phone
               case .youtube(_): return link.platform == SharedSocialLink.Platform.youtube
               case .tiktok(_): return link.platform == SharedSocialLink.Platform.tiktok
               }
           }
           
           return emptyLink.changeUrl(URL(string: link?.url ?? ""))
       }
    }
}

File: shared/src/commonMain/kotlin/com/tomorrow/convenire/shared/data/data_source/mapper/SocialLinkMapper.kt
package com.tomorrow.convenire.shared.data.data_source.mapper

import com.tomorrow.convenire.shared.data.data_source.model.SocialLinkDTO
import com.tomorrow.convenire.shared.data.data_source.utils.EntityMapper
import com.tomorrow.convenire.shared.domain.model.SocialLink

class SocialLinkMapper : EntityMapper<SocialLink, SocialLinkDTO> {
    override fun mapFromEntity(entity: SocialLinkDTO) = SocialLink(
        url = entity.url, platform = when (entity.platform) {
            SocialLinkDTO.Platform.Instagram -> SocialLink.Platform.Instagram
            SocialLinkDTO.Platform.Website -> SocialLink.Platform.Website
            SocialLinkDTO.Platform.LinkedIn -> SocialLink.Platform.LinkedIn
            SocialLinkDTO.Platform.Facebook -> SocialLink.Platform.Facebook
            SocialLinkDTO.Platform.Tiktok -> SocialLink.Platform.Tiktok
            SocialLinkDTO.Platform.Twitter -> SocialLink.Platform.Twitter
            SocialLinkDTO.Platform.Youtube -> SocialLink.Platform.Youtube
            SocialLinkDTO.Platform.WhatsApp -> SocialLink.Platform.WhatsApp
            SocialLinkDTO.Platform.Phone -> SocialLink.Platform.Phone
        }
    )

    override fun mapToEntity(domainModel: SocialLink): SocialLinkDTO {
        TODO("Not yet implemented")
    }
}